### PR TITLE
WorkerPool: protect against batch size 0

### PR DESCRIPTION
**Pull request by larham**
As part of a previous commit, the WorkerPool of threads will raise if
the number of workers is set to 0. This prevents a coding error from
resulting in no work, where work was expected.

This commit reaches through code to protect against the numWorkers
parameter being 0.

In several cases, the number of workers is set to the number of segments
or the number of databases. We do not protect those with the expectation
that something is more significantly wrong if those sums are 0, and that
an exception would be fitting in those cases.

This commit follows and is related to #3036 #3085

### CODE REVIEW

File: gpMgmt/bin/analyzedb [hunk 1]
```diff
@@ -355,9 +355,8 @@ class AnalyzeDb(Operation):
             if not self.silent and not userinput.ask_yesno(None, "\nContinue with Analyze", 'N'):
                 raise UserAbortedException()
 
-            num_workers = min(self.parallel_level, len(ordered_candidates))
-            logger.info("Starting analyze with %d workers..." % num_workers)
-            pool = AnalyzeWorkerPool(numWorkers=num_workers)
```
Comment (khuddlefish) [inline gpMgmt/bin/analyzedb]: Why did you make this change? If ordered_candidates < parallel_level you will now have extra workers in the pool. Also parallel_level is 5 by default and there is already a check protecting it from being < 1

File: gpMgmt/bin/gppylib/operations/utils.py [hunk 2]
```diff
@@ -54,23 +54,27 @@ def execute(self):
...
-    def execute(self):        
-        """TODO: Make Command a subclass of Operation. Then, make WorkerPool work with Operation objects."""
+
+    def execute(self):
+        if not self.operations or len(self.operations) == 0:
+            return
         pool = OperationWorkerPool(numWorkers=self.parallelism, operations=self.operations)
         pool.join()
         pool.haltWork()
-        return None
```
Comment (khuddlefish) [inline gpMgmt/bin/gppylib/operations/utils.py]: Is it still returning None like the comment claims?

Comment (larham) [reply gpMgmt/bin/gppylib/operations/utils.py]: Yes, I believe that every python method that has no "return" statement returns "None":

```
$ python
Python 2.7.13 (default, Apr  4 2017, 08:47:57)
>>> def foo():
...     x=1
...
>>> y=foo()
>>> print y
None
```

File: gpMgmt/bin/analyzedb (hunk 1 above)
Comment (larham) [reply gpMgmt/bin/analyzedb]: Yes, good point. The optimization you cite, of having the minimum number of threads, is lost with this change.  However, this was done with a reason:  we reasoned that the optimization as *less* important than the protection against a raised exception, and readability.  In particular, it is possible for Analyze to come up with 0 things to do. It seems nicest to do nothing in that case, without complaining. However, there has to be a thread in the pool for the

Comment (khuddlefish) [reply gpMgmt/bin/analyzedb]: Ok. I wasn't sure if it was possible for ordered_candidates to be empty or not.

**Pull request merged**